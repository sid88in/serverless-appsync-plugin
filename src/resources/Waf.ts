import { isEmpty, reduce } from 'lodash';
import {
  CfnResources,
  CfnWafAction,
  CfnWafRule,
  CfnWafRuleStatement,
} from '../types/cloudFormation';
import {
  ApiKeyConfig,
  WafConfig,
  WafRule,
  WafRuleAction,
  WafRuleDisableIntrospection,
  WafThrottleConfig,
} from '../types/plugin';
import { Api } from './Api';
import { toCfnKeys } from '../utils';

export class Waf {
  constructor(private api: Api, private config: WafConfig) {}

  compile(): CfnResources {
    const wafConfig = this.config;
    if (wafConfig.enabled === false) {
      return {};
    }
    const apiLogicalId = this.api.naming.getApiLogicalId();
    const wafAssocLogicalId = this.api.naming.getWafAssociationLogicalId();

    if (wafConfig.arn) {
      return {
        [wafAssocLogicalId]: {
          Type: 'AWS::WAFv2::WebACLAssociation',
          Properties: {
            ResourceArn: { 'Fn::GetAtt': [apiLogicalId, 'Arn'] },
            WebACLArn: wafConfig.arn,
          },
        },
      };
    }

    const name = wafConfig.name || `${this.api.config.name}Waf`;
    const wafLogicalId = this.api.naming.getWafLogicalId();
    const defaultActionSource = wafConfig.defaultAction || 'Allow';
    const defaultAction: CfnWafAction =
      typeof defaultActionSource === 'string'
        ? { [defaultActionSource]: {} }
        : defaultActionSource;

    return {
      [wafLogicalId]: {
        Type: 'AWS::WAFv2::WebACL',
        Properties: {
          DefaultAction: defaultAction,
          Scope: 'REGIONAL',
          Description:
            wafConfig.description ||
            `ACL rules for AppSync ${this.api.config.name}`,
          Name: name,
          Rules: this.buildWafRules(),
          VisibilityConfig: this.getWafVisibilityConfig(
            this.config.visibilityConfig,
            name,
          ),
          Tags: this.api.getTagsConfig(),
        },
      },
      [wafAssocLogicalId]: {
        Type: 'AWS::WAFv2::WebACLAssociation',
        Properties: {
          ResourceArn: { 'Fn::GetAtt': [apiLogicalId, 'Arn'] },
          WebACLArn: { 'Fn::GetAtt': [wafLogicalId, 'Arn'] },
        },
      },
    };
  }

  buildWafRules() {
    const rules = this.config.rules || [];

    let defaultPriority = 100;
    return rules
      .map((rule) => this.buildWafRule(rule))
      .concat(this.buildApiKeysWafRules())
      .map((rule) => ({
        ...rule,
        Priority: rule.Priority || defaultPriority++,
      }));
  }

  buildWafRule(rule: WafRule, defaultNamePrefix?: string): CfnWafRule {
    // Throttle pre-set rule
    if (rule === 'throttle') {
      return this.buildThrottleRule({}, defaultNamePrefix);
    } else if (typeof rule === 'object' && 'throttle' in rule) {
      return this.buildThrottleRule(rule.throttle, defaultNamePrefix);
    }

    // Disable Introspection pre-set rule
    if (rule === 'disableIntrospection') {
      return this.buildDisableIntrospectionRule({}, defaultNamePrefix);
    } else if ('disableIntrospection' in rule) {
      return this.buildDisableIntrospectionRule(
        rule.disableIntrospection,
        defaultNamePrefix,
      );
    }

    const action: WafRuleAction = rule.action || 'Allow';
    const overrideAction = rule.overrideAction;

    const result: CfnWafRule = {
      Name: rule.name,
      Priority: rule.priority,
      Statement: rule.statement,
      VisibilityConfig: this.getWafVisibilityConfig(
        rule.visibilityConfig,
        rule.name,
      ),
    };

    if (overrideAction) {
      result.OverrideAction = toCfnKeys(overrideAction);
    } else {
      result.Action = { [action]: {} };
    }

    return result;
  }

  buildApiKeysWafRules(): CfnWafRule[] {
    return (
      reduce(
        this.api.config.apiKeys,
        (rules, key) => rules.concat(this.buildApiKeyRules(key)),
        [] as CfnWafRule[],
      ) || []
    );
  }

  buildApiKeyRules(key: ApiKeyConfig) {
    const rules = key.wafRules;
    // Build the rule and add a matching rule for the X-Api-Key header
    // for the given api key
    const allRules: CfnWafRule[] = [];
    rules?.forEach((keyRule) => {
      const builtRule = this.buildWafRule(keyRule, key.name);
      const logicalIdApiKey = this.api.naming.getApiKeyLogicalId(key.name);
      const { Statement: baseStatement } = builtRule;
      const apiKeyStatement: CfnWafRuleStatement = {
        ByteMatchStatement: {
          FieldToMatch: {
            SingleHeader: { Name: 'X-Api-key' },
          },
          PositionalConstraint: 'EXACTLY',
          SearchString: { 'Fn::GetAtt': [logicalIdApiKey, 'ApiKey'] },
          TextTransformations: [
            {
              Type: 'LOWERCASE',
              Priority: 0,
            },
          ],
        },
      };

      let statement: CfnWafRuleStatement;
      if (baseStatement && baseStatement?.RateBasedStatement) {
        let ScopeDownStatement: CfnWafRuleStatement;
        // For RateBasedStatement, use the api rule as ScopeDownStatement
        // merge if with existing needed
        if (baseStatement.RateBasedStatement?.ScopeDownStatement) {
          ScopeDownStatement = this.mergeWafRuleStatements([
            baseStatement.RateBasedStatement.ScopeDownStatement,
            apiKeyStatement,
          ]);
        } else {
          ScopeDownStatement = apiKeyStatement;
        }
        // RateBasedStatement
        statement = {
          RateBasedStatement: {
            ...baseStatement.RateBasedStatement,
            ScopeDownStatement,
          },
        };
      } else if (!isEmpty(baseStatement)) {
        // Other rules: combine them (And Statement)
        statement = this.mergeWafRuleStatements([
          baseStatement,
          apiKeyStatement,
        ]);
      } else {
        // No statement, the rule is the API key rule itself
        statement = apiKeyStatement;
      }

      allRules.push({
        ...builtRule,
        Statement: statement,
      });
    });

    return allRules;
  }

  mergeWafRuleStatements(
    statements: CfnWafRuleStatement[],
  ): CfnWafRuleStatement {
    return {
      AndStatement: {
        Statements: statements,
      },
    };
  }

  getWafVisibilityConfig(
    visibilityConfig: Record<string, unknown> | undefined = {},
    defaultName: string,
  ) {
    return {
      CloudWatchMetricsEnabled:
        visibilityConfig.cloudWatchMetricsEnabled ??
        this.config.visibilityConfig?.cloudWatchMetricsEnabled ??
        true,
      MetricName: visibilityConfig.name || defaultName,
      SampledRequestsEnabled:
        visibilityConfig.sampledRequestsEnabled ??
        this.config.visibilityConfig?.sampledRequestsEnabled ??
        true,
    };
  }

  buildDisableIntrospectionRule(
    config: WafRuleDisableIntrospection['disableIntrospection'],
    defaultNamePrefix?: string,
  ): CfnWafRule {
    const Name =
      config.name || `${defaultNamePrefix || ''}DisableIntrospection`;

    return {
      Action: {
        Block: {},
      },
      Name,
      Priority: config.priority,
      Statement: {
        OrStatement: {
          Statements: [
            {
              // Block all requests > 8kb
              // https://docs.aws.amazon.com/waf/latest/developerguide/web-request-body-inspection.html
              SizeConstraintStatement: {
                ComparisonOperator: 'GT',
                FieldToMatch: {
                  Body: {},
                },
                Size: 8 * 1024,
                TextTransformations: [
                  {
                    Type: 'NONE',
                    Priority: 0,
                  },
                ],
              },
            },
            {
              ByteMatchStatement: {
                FieldToMatch: {
                  Body: {},
                },
                PositionalConstraint: 'CONTAINS',
                SearchString: '__schema',
                TextTransformations: [
                  {
                    Type: 'COMPRESS_WHITE_SPACE',
                    Priority: 0,
                  },
                ],
              },
            },
          ],
        },
      },
      VisibilityConfig: this.getWafVisibilityConfig(
        typeof config === 'object' ? config.visibilityConfig : undefined,
        Name,
      ),
    };
  }

  buildThrottleRule(
    config: WafThrottleConfig,
    defaultNamePrefix?: string,
  ): CfnWafRule {
    let Name = `${defaultNamePrefix || ''}Throttle`;
    let Limit = 100;
    let AggregateKeyType = 'IP';
    let ForwardedIPConfig;
    let Priority;
    let ScopeDownStatement;

    if (typeof config === 'number') {
      Limit = config;
    } else if (typeof config === 'object') {
      Name = config.name || Name;
      AggregateKeyType = config.aggregateKeyType || AggregateKeyType;
      Limit = config.limit || Limit;
      Priority = config.priority;
      ScopeDownStatement = config.scopeDownStatement;
      if (AggregateKeyType === 'FORWARDED_IP') {
        ForwardedIPConfig = {
          HeaderName: config.forwardedIPConfig?.headerName || 'X-Forwarded-For',
          FallbackBehavior:
            config.forwardedIPConfig?.fallbackBehavior || 'MATCH',
        };
      }
    }

    return {
      Action: {
        Block: {},
      },
      Name,
      Priority,
      Statement: {
        RateBasedStatement: {
          AggregateKeyType,
          Limit,
          ForwardedIPConfig,
          ScopeDownStatement,
        },
      },
      VisibilityConfig: this.getWafVisibilityConfig(
        typeof config === 'object' ? config.visibilityConfig : undefined,
        Name,
      ),
    };
  }
}
