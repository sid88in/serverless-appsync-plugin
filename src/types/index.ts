import { BuildOptions } from 'esbuild';
import {
  ApiKeyConfig,
  Auth,
  Substitutions,
  CachingConfig,
  DomainConfig,
  LoggingConfig,
  WafConfig,
  DsDynamoDBConfig,
  DsEventBridgeConfig,
  DsHttpConfig,
  DsLambdaConfig,
  DsNone,
  DsOpenSearchConfig,
  DsRelationalDbConfig,
  SyncConfig,
  EnvironmentVariables,
} from './common';
export * from './common';

export type AppSyncConfig = {
  name: string;
  schema?: string | string[];
  authentication: Auth;
  additionalAuthentications?: Auth[];
  domain?: DomainConfig;
  apiKeys?: (ApiKeyConfig | string)[];
  resolvers?: Record<string, ResolverConfig>[] | Record<string, ResolverConfig>;
  pipelineFunctions?:
    | Record<string, PipelineFunctionConfig>[]
    | Record<string, PipelineFunctionConfig>;
  dataSources:
    | Record<string, DataSourceConfig>[]
    | Record<string, DataSourceConfig>;
  substitutions?: Substitutions;
  environment?: EnvironmentVariables;
  xrayEnabled?: boolean;
  logging?: LoggingConfig;
  caching?: CachingConfig;
  waf?: WafConfig;
  tags?: Record<string, string>;
  visibility?: 'GLOBAL' | 'PRIVATE';
  esbuild?: BuildOptions | false;
  introspection?: boolean;
  queryDepthLimit?: number;
  resolverCountLimit?: number;
};

export type BaseResolverConfig = {
  field?: string;
  type?: string;
  request?: string | false;
  response?: string | false;
  code?: string;
  caching?:
    | {
        ttl?: number;
        keys?: string[];
      }
    | boolean;
  sync?: SyncConfig;
  substitutions?: Substitutions;
};

export type ResolverConfig = UnitResolverConfig | PipelineResolverConfig;

export type UnitResolverConfig = BaseResolverConfig & {
  kind: 'UNIT';
  dataSource: string | DataSourceConfig;
  maxBatchSize?: number;
};

export type PipelineResolverConfig = BaseResolverConfig & {
  kind?: 'PIPELINE';
  functions: (string | PipelineFunctionConfig)[];
};

export type DataSourceConfig = {
  description?: string;
} & (
  | DsHttpConfig
  | DsDynamoDBConfig
  | DsRelationalDbConfig
  | DsOpenSearchConfig
  | DsLambdaConfig
  | DsEventBridgeConfig
  | DsNone
);

export type PipelineFunctionConfig = {
  dataSource: string | DataSourceConfig;
  description?: string;
  code?: string;
  request?: string;
  response?: string;
  maxBatchSize?: number;
  substitutions?: Substitutions;
  sync?: SyncConfig;
};
