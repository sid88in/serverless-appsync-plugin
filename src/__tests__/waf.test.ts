import { Api } from '../resources/Api';
import { ApiKeyConfig, WafRule } from '../types/plugin';
import { each } from 'lodash';
import { Waf } from '../resources/Waf';
import * as given from './given';

const plugin = given.plugin();

describe('Waf', () => {
  describe('Base Resources', () => {
    it('should generate waf Resources', () => {
      const api = new Api(given.appSyncConfig(), plugin);
      const waf = new Waf(api, {
        enabled: true,
        name: 'Waf',
        defaultAction: 'Allow',
        description: 'My Waf ACL',
        visibilityConfig: {
          cloudWatchMetricsEnabled: true,
          name: 'MyVisibilityConfig',
          sampledRequestsEnabled: true,
        },
        rules: [],
      });
      expect(waf.compile()).toMatchSnapshot();
    });

    it('should generate waf Resources without tags', () => {
      const api = new Api(
        given.appSyncConfig({
          tags: undefined,
        }),
        plugin,
      );
      const waf = new Waf(api, {
        enabled: true,
        name: 'Waf',
        defaultAction: 'Allow',
        description: 'My Waf ACL',
        visibilityConfig: {
          cloudWatchMetricsEnabled: true,
          name: 'MyVisibilityConfig',
          sampledRequestsEnabled: true,
        },
        rules: [],
      });
      expect(waf.compile()).toMatchSnapshot();
    });

    it('should not generate waf Resources if disabled', () => {
      const api = new Api(
        given.appSyncConfig({
          waf: {
            enabled: false,
            name: 'Waf',
            rules: [],
          },
        }),
        plugin,
      );
      expect(api.compileWafRules()).toEqual({});
    });

    it('should generate only the waf association', () => {
      const api = new Api(given.appSyncConfig(), plugin);
      const waf = new Waf(api, {
        enabled: true,
        arn: 'arn:aws:wafv2:us-east-1:123456789012:regional/webacl/my-Waf/d7b694d2-4f7d-4dd6-a9a9-843dd1931330',
      });
      expect(waf.compile()).toMatchSnapshot();
    });
  });

  describe('Throttle rules', () => {
    const api = new Api(given.appSyncConfig(), plugin);
    const waf = new Waf(api, {
      name: 'Waf',
      rules: [],
    });

    it('should generate a preset rule', () => {
      expect(waf.buildWafRule('throttle', 'Base')).toMatchSnapshot();
    });

    it('should generate a preset rule with limit', () => {
      expect(waf.buildWafRule({ throttle: 500 }, 'Base')).toMatchSnapshot();
    });

    it('should generate a preset rule with config', () => {
      expect(
        waf.buildWafRule(
          {
            throttle: {
              priority: 300,
              limit: 200,
              aggregateKeyType: 'FORWARDED_IP',
              forwardedIPConfig: {
                headerName: 'X-Forwarded-To',
                fallbackBehavior: 'MATCH',
              },
              visibilityConfig: {
                name: 'ThrottleRule',
                cloudWatchMetricsEnabled: false,
                sampledRequestsEnabled: false,
              },
            },
          },

          'Base',
        ),
      ).toMatchSnapshot();
    });
  });

  describe('Disable introspection', () => {
    const api = new Api(given.appSyncConfig(), plugin);
    const waf = new Waf(api, {
      name: 'Waf',
      rules: [],
    });

    it('should generate a preset rule', () => {
      expect(
        waf.buildWafRule('disableIntrospection', 'Base'),
      ).toMatchSnapshot();
    });

    it('should generate a preset rule with custon config', () => {
      expect(
        waf.buildWafRule(
          {
            disableIntrospection: {
              priority: 200,
              visibilityConfig: {
                name: 'DisableIntrospection',
                sampledRequestsEnabled: false,
                cloudWatchMetricsEnabled: false,
              },
            },
          },
          'Base',
        ),
      ).toMatchSnapshot();
    });
  });

  describe('Custom rules', () => {
    const api = new Api(given.appSyncConfig(), plugin);
    const waf = new Waf(api, {
      name: 'Waf',
      rules: [],
    });

    it('should generate a custom rule', () => {
      expect(
        waf.buildWafRule(
          {
            name: 'disable US',
            priority: 200,
            action: 'Block',
            statement: {
              GeoMatchStatement: {
                CountryCodes: ['US'],
              },
            },
          },
          'Base',
        ),
      ).toMatchSnapshot();
    });

    it('should generate a custom rule with ManagedRuleGroup', () => {
      expect(
        waf.buildWafRule(
          {
            name: 'MyRule1',
            priority: 200,
            overrideAction: {
              None: {},
            },
            statement: {
              ManagedRuleGroupStatement: {
                Name: 'AWSManagedRulesCommonRuleSet',
                VendorName: 'AWS',
              },
            },
          },
          'Base',
        ),
      ).toMatchSnapshot();
    });
  });

  describe('ApiKey rules', () => {
    const configs: Record<string, WafRule> = {
      throttle: 'throttle',
      disableIntrospection: 'disableIntrospection',
      customRule: {
        name: 'MyCustomRule',
        statement: {
          GeoMatchStatement: {
            CountryCodes: ['US'],
          },
        },
      },
      throttleWithStatements: {
        throttle: {
          name: 'Throttle rule with custom ScopeDownStatement',
          limit: 100,
          scopeDownStatement: {
            ByteMatchStatement: {
              FieldToMatch: {
                SingleHeader: { Name: 'X-Foo' },
              },
              PositionalConstraint: 'EXACTLY',
              SearchString: 'Bar',
              TextTransformations: [
                {
                  Type: 'LOWERCASE',
                  Priority: 0,
                },
              ],
            },
          },
        },
      },
      emptyStatements: {
        name: 'rulesWithoutStatements',
        statement: {},
      },
    };
    const api = new Api(given.appSyncConfig(), plugin);
    const waf = new Waf(api, {
      name: 'Waf',
      rules: [],
    });

    each(configs, (rule, name) => {
      it(`should generate a rule for ${name}`, () => {
        const apiConfig: ApiKeyConfig = {
          name: 'MyKey',
          wafRules: [rule],
        };
        expect(waf.buildApiKeyRules(apiConfig)).toMatchSnapshot();
      });
    });
  });
});
