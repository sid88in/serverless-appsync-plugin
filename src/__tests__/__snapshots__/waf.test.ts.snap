// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Waf ApiKey rules should generate a rule for customRule 1`] = `
Array [
  Object {
    "Action": Object {
      "Allow": Object {},
    },
    "Name": "MyCustomRule",
    "Priority": undefined,
    "Statement": Object {
      "AndStatement": Object {
        "Statements": Array [
          Object {
            "GeoMatchStatement": Object {
              "CountryCodes": Array [
                "US",
              ],
            },
          },
          Object {
            "ByteMatchStatement": Object {
              "FieldToMatch": Object {
                "SingleHeader": Object {
                  "Name": "X-Api-key",
                },
              },
              "PositionalConstraint": "EXACTLY",
              "SearchString": Object {
                "Fn::GetAtt": Array [
                  "GraphQlApiMyKey",
                  "ApiKey",
                ],
              },
              "TextTransformations": Array [
                Object {
                  "Priority": 0,
                  "Type": "LOWERCASE",
                },
              ],
            },
          },
        ],
      },
    },
    "VisibilityConfig": Object {
      "CloudWatchMetricsEnabled": true,
      "MetricName": "MyCustomRule",
      "SampledRequestsEnabled": true,
    },
  },
]
`;

exports[`Waf ApiKey rules should generate a rule for disableIntrospection 1`] = `
Array [
  Object {
    "Action": Object {
      "Block": Object {},
    },
    "Name": "MyKeyDisableIntrospection",
    "Priority": undefined,
    "Statement": Object {
      "AndStatement": Object {
        "Statements": Array [
          Object {
            "OrStatement": Object {
              "Statements": Array [
                Object {
                  "SizeConstraintStatement": Object {
                    "ComparisonOperator": "GT",
                    "FieldToMatch": Object {
                      "Body": Object {},
                    },
                    "Size": 8192,
                    "TextTransformations": Array [
                      Object {
                        "Priority": 0,
                        "Type": "NONE",
                      },
                    ],
                  },
                },
                Object {
                  "ByteMatchStatement": Object {
                    "FieldToMatch": Object {
                      "Body": Object {},
                    },
                    "PositionalConstraint": "CONTAINS",
                    "SearchString": "__schema",
                    "TextTransformations": Array [
                      Object {
                        "Priority": 0,
                        "Type": "COMPRESS_WHITE_SPACE",
                      },
                    ],
                  },
                },
              ],
            },
          },
          Object {
            "ByteMatchStatement": Object {
              "FieldToMatch": Object {
                "SingleHeader": Object {
                  "Name": "X-Api-key",
                },
              },
              "PositionalConstraint": "EXACTLY",
              "SearchString": Object {
                "Fn::GetAtt": Array [
                  "GraphQlApiMyKey",
                  "ApiKey",
                ],
              },
              "TextTransformations": Array [
                Object {
                  "Priority": 0,
                  "Type": "LOWERCASE",
                },
              ],
            },
          },
        ],
      },
    },
    "VisibilityConfig": Object {
      "CloudWatchMetricsEnabled": true,
      "MetricName": "MyKeyDisableIntrospection",
      "SampledRequestsEnabled": true,
    },
  },
]
`;

exports[`Waf ApiKey rules should generate a rule for emptyStatements 1`] = `
Array [
  Object {
    "Action": Object {
      "Allow": Object {},
    },
    "Name": "rulesWithoutStatements",
    "Priority": undefined,
    "Statement": Object {
      "ByteMatchStatement": Object {
        "FieldToMatch": Object {
          "SingleHeader": Object {
            "Name": "X-Api-key",
          },
        },
        "PositionalConstraint": "EXACTLY",
        "SearchString": Object {
          "Fn::GetAtt": Array [
            "GraphQlApiMyKey",
            "ApiKey",
          ],
        },
        "TextTransformations": Array [
          Object {
            "Priority": 0,
            "Type": "LOWERCASE",
          },
        ],
      },
    },
    "VisibilityConfig": Object {
      "CloudWatchMetricsEnabled": true,
      "MetricName": "rulesWithoutStatements",
      "SampledRequestsEnabled": true,
    },
  },
]
`;

exports[`Waf ApiKey rules should generate a rule for throttle 1`] = `
Array [
  Object {
    "Action": Object {
      "Block": Object {},
    },
    "Name": "MyKeyThrottle",
    "Priority": undefined,
    "Statement": Object {
      "RateBasedStatement": Object {
        "AggregateKeyType": "IP",
        "ForwardedIPConfig": undefined,
        "Limit": 100,
        "ScopeDownStatement": Object {
          "ByteMatchStatement": Object {
            "FieldToMatch": Object {
              "SingleHeader": Object {
                "Name": "X-Api-key",
              },
            },
            "PositionalConstraint": "EXACTLY",
            "SearchString": Object {
              "Fn::GetAtt": Array [
                "GraphQlApiMyKey",
                "ApiKey",
              ],
            },
            "TextTransformations": Array [
              Object {
                "Priority": 0,
                "Type": "LOWERCASE",
              },
            ],
          },
        },
      },
    },
    "VisibilityConfig": Object {
      "CloudWatchMetricsEnabled": true,
      "MetricName": "MyKeyThrottle",
      "SampledRequestsEnabled": true,
    },
  },
]
`;

exports[`Waf ApiKey rules should generate a rule for throttleWithStatements 1`] = `
Array [
  Object {
    "Action": Object {
      "Block": Object {},
    },
    "Name": "Throttle rule with custom ScopeDownStatement",
    "Priority": undefined,
    "Statement": Object {
      "RateBasedStatement": Object {
        "AggregateKeyType": "IP",
        "ForwardedIPConfig": undefined,
        "Limit": 100,
        "ScopeDownStatement": Object {
          "AndStatement": Object {
            "Statements": Array [
              Object {
                "ByteMatchStatement": Object {
                  "FieldToMatch": Object {
                    "SingleHeader": Object {
                      "Name": "X-Foo",
                    },
                  },
                  "PositionalConstraint": "EXACTLY",
                  "SearchString": "Bar",
                  "TextTransformations": Array [
                    Object {
                      "Priority": 0,
                      "Type": "LOWERCASE",
                    },
                  ],
                },
              },
              Object {
                "ByteMatchStatement": Object {
                  "FieldToMatch": Object {
                    "SingleHeader": Object {
                      "Name": "X-Api-key",
                    },
                  },
                  "PositionalConstraint": "EXACTLY",
                  "SearchString": Object {
                    "Fn::GetAtt": Array [
                      "GraphQlApiMyKey",
                      "ApiKey",
                    ],
                  },
                  "TextTransformations": Array [
                    Object {
                      "Priority": 0,
                      "Type": "LOWERCASE",
                    },
                  ],
                },
              },
            ],
          },
        },
      },
    },
    "VisibilityConfig": Object {
      "CloudWatchMetricsEnabled": true,
      "MetricName": "Throttle rule with custom ScopeDownStatement",
      "SampledRequestsEnabled": true,
    },
  },
]
`;

exports[`Waf Base Resources should generate only the waf association 1`] = `
Object {
  "GraphQlWafAssoc": Object {
    "Properties": Object {
      "ResourceArn": Object {
        "Fn::GetAtt": Array [
          "GraphQlApi",
          "Arn",
        ],
      },
      "WebACLArn": "arn:aws:wafv2:us-east-1:123456789012:regional/webacl/my-Waf/d7b694d2-4f7d-4dd6-a9a9-843dd1931330",
    },
    "Type": "AWS::WAFv2::WebACLAssociation",
  },
}
`;

exports[`Waf Base Resources should generate waf Resources 1`] = `
Object {
  "GraphQlWaf": Object {
    "Properties": Object {
      "DefaultAction": Object {
        "Allow": Object {},
      },
      "Description": "My Waf ACL",
      "Name": "Waf",
      "Rules": Array [],
      "Scope": "REGIONAL",
      "Tags": Array [
        Object {
          "Key": "stage",
          "Value": "Dev",
        },
      ],
      "VisibilityConfig": Object {
        "CloudWatchMetricsEnabled": true,
        "MetricName": "MyVisibilityConfig",
        "SampledRequestsEnabled": true,
      },
    },
    "Type": "AWS::WAFv2::WebACL",
  },
  "GraphQlWafAssoc": Object {
    "Properties": Object {
      "ResourceArn": Object {
        "Fn::GetAtt": Array [
          "GraphQlApi",
          "Arn",
        ],
      },
      "WebACLArn": Object {
        "Fn::GetAtt": Array [
          "GraphQlWaf",
          "Arn",
        ],
      },
    },
    "Type": "AWS::WAFv2::WebACLAssociation",
  },
}
`;

exports[`Waf Base Resources should generate waf Resources without tags 1`] = `
Object {
  "GraphQlWaf": Object {
    "Properties": Object {
      "DefaultAction": Object {
        "Allow": Object {},
      },
      "Description": "My Waf ACL",
      "Name": "Waf",
      "Rules": Array [],
      "Scope": "REGIONAL",
      "Tags": undefined,
      "VisibilityConfig": Object {
        "CloudWatchMetricsEnabled": true,
        "MetricName": "MyVisibilityConfig",
        "SampledRequestsEnabled": true,
      },
    },
    "Type": "AWS::WAFv2::WebACL",
  },
  "GraphQlWafAssoc": Object {
    "Properties": Object {
      "ResourceArn": Object {
        "Fn::GetAtt": Array [
          "GraphQlApi",
          "Arn",
        ],
      },
      "WebACLArn": Object {
        "Fn::GetAtt": Array [
          "GraphQlWaf",
          "Arn",
        ],
      },
    },
    "Type": "AWS::WAFv2::WebACLAssociation",
  },
}
`;

exports[`Waf Custom rules should generate a custom rule 1`] = `
Object {
  "Action": Object {
    "Block": Object {},
  },
  "Name": "disable US",
  "Priority": 200,
  "Statement": Object {
    "GeoMatchStatement": Object {
      "CountryCodes": Array [
        "US",
      ],
    },
  },
  "VisibilityConfig": Object {
    "CloudWatchMetricsEnabled": true,
    "MetricName": "disable US",
    "SampledRequestsEnabled": true,
  },
}
`;

exports[`Waf Custom rules should generate a custom rule with ManagedRuleGroup 1`] = `
Object {
  "Name": "MyRule1",
  "OverrideAction": Object {
    "None": Object {},
  },
  "Priority": 200,
  "Statement": Object {
    "ManagedRuleGroupStatement": Object {
      "Name": "AWSManagedRulesCommonRuleSet",
      "VendorName": "AWS",
    },
  },
  "VisibilityConfig": Object {
    "CloudWatchMetricsEnabled": true,
    "MetricName": "MyRule1",
    "SampledRequestsEnabled": true,
  },
}
`;

exports[`Waf Disable introspection should generate a preset rule 1`] = `
Object {
  "Action": Object {
    "Block": Object {},
  },
  "Name": "BaseDisableIntrospection",
  "Priority": undefined,
  "Statement": Object {
    "OrStatement": Object {
      "Statements": Array [
        Object {
          "SizeConstraintStatement": Object {
            "ComparisonOperator": "GT",
            "FieldToMatch": Object {
              "Body": Object {},
            },
            "Size": 8192,
            "TextTransformations": Array [
              Object {
                "Priority": 0,
                "Type": "NONE",
              },
            ],
          },
        },
        Object {
          "ByteMatchStatement": Object {
            "FieldToMatch": Object {
              "Body": Object {},
            },
            "PositionalConstraint": "CONTAINS",
            "SearchString": "__schema",
            "TextTransformations": Array [
              Object {
                "Priority": 0,
                "Type": "COMPRESS_WHITE_SPACE",
              },
            ],
          },
        },
      ],
    },
  },
  "VisibilityConfig": Object {
    "CloudWatchMetricsEnabled": true,
    "MetricName": "BaseDisableIntrospection",
    "SampledRequestsEnabled": true,
  },
}
`;

exports[`Waf Disable introspection should generate a preset rule with custon config 1`] = `
Object {
  "Action": Object {
    "Block": Object {},
  },
  "Name": "BaseDisableIntrospection",
  "Priority": 200,
  "Statement": Object {
    "OrStatement": Object {
      "Statements": Array [
        Object {
          "SizeConstraintStatement": Object {
            "ComparisonOperator": "GT",
            "FieldToMatch": Object {
              "Body": Object {},
            },
            "Size": 8192,
            "TextTransformations": Array [
              Object {
                "Priority": 0,
                "Type": "NONE",
              },
            ],
          },
        },
        Object {
          "ByteMatchStatement": Object {
            "FieldToMatch": Object {
              "Body": Object {},
            },
            "PositionalConstraint": "CONTAINS",
            "SearchString": "__schema",
            "TextTransformations": Array [
              Object {
                "Priority": 0,
                "Type": "COMPRESS_WHITE_SPACE",
              },
            ],
          },
        },
      ],
    },
  },
  "VisibilityConfig": Object {
    "CloudWatchMetricsEnabled": false,
    "MetricName": "DisableIntrospection",
    "SampledRequestsEnabled": false,
  },
}
`;

exports[`Waf Throttle rules should generate a preset rule 1`] = `
Object {
  "Action": Object {
    "Block": Object {},
  },
  "Name": "BaseThrottle",
  "Priority": undefined,
  "Statement": Object {
    "RateBasedStatement": Object {
      "AggregateKeyType": "IP",
      "ForwardedIPConfig": undefined,
      "Limit": 100,
      "ScopeDownStatement": undefined,
    },
  },
  "VisibilityConfig": Object {
    "CloudWatchMetricsEnabled": true,
    "MetricName": "BaseThrottle",
    "SampledRequestsEnabled": true,
  },
}
`;

exports[`Waf Throttle rules should generate a preset rule with config 1`] = `
Object {
  "Action": Object {
    "Block": Object {},
  },
  "Name": "BaseThrottle",
  "Priority": 300,
  "Statement": Object {
    "RateBasedStatement": Object {
      "AggregateKeyType": "FORWARDED_IP",
      "ForwardedIPConfig": Object {
        "FallbackBehavior": "MATCH",
        "HeaderName": "X-Forwarded-To",
      },
      "Limit": 200,
      "ScopeDownStatement": undefined,
    },
  },
  "VisibilityConfig": Object {
    "CloudWatchMetricsEnabled": false,
    "MetricName": "ThrottleRule",
    "SampledRequestsEnabled": false,
  },
}
`;

exports[`Waf Throttle rules should generate a preset rule with limit 1`] = `
Object {
  "Action": Object {
    "Block": Object {},
  },
  "Name": "BaseThrottle",
  "Priority": undefined,
  "Statement": Object {
    "RateBasedStatement": Object {
      "AggregateKeyType": "IP",
      "ForwardedIPConfig": undefined,
      "Limit": 500,
      "ScopeDownStatement": undefined,
    },
  },
  "VisibilityConfig": Object {
    "CloudWatchMetricsEnabled": true,
    "MetricName": "BaseThrottle",
    "SampledRequestsEnabled": true,
  },
}
`;
